File: src/2024/day3.zig
const std = @import("std");
const utils = @import("utils");

pub fn main() void {
    // https://adventofcode.com/2024/day/3
    utils.printTitle(2024, 3, "Mull It Over");

    const m = utils.Monitor.init();
    defer m.deinit();
    const input = comptime utils.trim_input(@embedFile("input"));

    @setEvalBranchQuota(80_000);
    const result = comptime run(input);
    utils.printAnswer(1, result.part1);
    comptime std.debug.assert(167090022 == result.part1);

    utils.printAnswer(2, result.part2);
    comptime std.debug.assert(89823704 == result.part2);
}

const State = enum {
    num1,
    num2,
    end,
};

const MulResult = struct {
    len: usize,
    product: i64,
};

fn is_valid(input: []const u8) ?MulResult {
    var num1: [3]u8 = undefined;
    var num1_len: usize = 0;

    var num2: [3]u8 = undefined;
    var num2_len: usize = 0;

    if (!std.mem.eql(u8, input[0..4], "mul(")) {
        return null;
    }

    var state = State.num1;
    var len: usize = 0;

    outer: for (4.., input[4..]) |i, c| {
        switch (state) {
            .num1 => {
                switch (c) {
                    ',' => state = .num2,
                    '0'...'9' => {
                        if (num1_len == 3) return null;
                        num1[num1_len] = c;
                        num1_len += 1;
                    },
                    else => return null,
                }
            },
            .num2 => {
                switch (c) {
                    ')' => {
                        state = .end;
                        len = i;
                        break :outer;
                    },
                    '0'...'9' => {
                        if (num2_len == 3) return null;
                        num2[num2_len] = c;
                        num2_len += 1;
                    },
                    else => return null,
                }
            },
            else => {},
        }
    }

    if (state != .end) {
        return null;
    }

    const a = std.fmt.parseInt(i64, num1[0..num1_len], 10) catch unreachable;
    const b = std.fmt.parseInt(i64, num2[0..num2_len], 10) catch unreachable;
    return .{ .len = len, .product = a * b };
}

fn run(input: []const u8) utils.Result(i64) {
    var skip: usize = 0;
    var result = utils.Result(i64).init();
    var enabled = true;

    for (0.., input) |i, c| {
        if (skip != 0) {
            skip -= 1;
            continue;
        }

        switch (c) {
            'd' => {
                const do = "do()";
                const dont = "don't()";
                if (std.mem.startsWith(u8, input[i..], do)) {
                    enabled = true;
                    skip = do.len - 1;
                } else if (std.mem.startsWith(u8, input[i..], dont)) {
                    enabled = false;
                    skip = dont.len - 1;
                }
            },
            'm' => {
                const mul_max_len = "mul(XXX,XXX)".len;
                const chunk = input[i..@min(input.len, i + mul_max_len)];
                if (is_valid(chunk)) |r| {
                    skip = r.len - 1;
                    result.part1 += r.product;

                    if (enabled) {
                        result.part2 += r.product;
                    }
                }
            },
            else => {},
        }
    }

    return result;
}

test "part 1" {
    const example =
        \\xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))
    ;
    comptime try std.testing.expectEqual(@as(i32, 161), run(example).part1);
}

test "part 2" {
    const example =
        \\xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))
    ;
    comptime try std.testing.expectEqual(48, run(example).part2);
}

File: src/2024/day3.rs
fn main() {
    // https://adventofcode.com/2024/day/3
    println!("Day 3, 2024: Mull It Over");

    debug_assert_eq!(161, run(get_example1()).0);
    debug_assert_eq!(48, run(get_example2()).1);

    let (p1, p2) = run(get_input());
    println!("  Part 1: {}", p1);
    assert_eq!(167090022, p1);

    println!("  Part 2: {}", p2);
    assert_eq!(89823704, p2);
}

fn run(input: &'static str) -> (u64, u64) {
    const WINDOW_SIZE: usize = "mul(xxx,xxx)".len();
    let re_mul = regex::Regex::new(r"^mul\((\d{1,3}),(\d{1,3})\)").unwrap();
    let mut part1: u64 = 0;
    let mut part2: u64 = 0;
    let mut enabled = true;

    for chunk in (input.to_owned() + "padpadpad")
        .as_bytes()
        .windows(WINDOW_SIZE)
    {
        let s = std::str::from_utf8(chunk).unwrap();

        if chunk.starts_with(b"do()") {
            enabled = true;
        } else if chunk.starts_with(b"don't()") {
            enabled = false;
        }

        if let Some(matches) = re_mul.captures(s) {
            let a: u64 = matches.get(1).unwrap().as_str().parse().unwrap();
            let b: u64 = matches.get(2).unwrap().as_str().parse().unwrap();
            part1 += a * b;

            if enabled {
                part2 += a * b;
            }
        }
    }

    (part1, part2)
}

fn get_input() -> &'static str {
    include_str!("../../inputs/2024/day3.txt").trim_end()
}

fn get_example1() -> &'static str {
    "xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))"
}

fn get_example2() -> &'static str {
    "xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))"
}
